docs/rfc/rfc-003-performance-questions.md

```diff
@@ -0,0 +1,270 @@
+P2P stack can scale to.
+
+We need to more concretely understand the source of issue and determine what layer
+is causing a problem. It's possible that the P2P layer, in the absence of any reactors
+sending data, is perfectly capable of managing thousands of peer connections. For
+a reasonable networking and application setup, thousands of connections should not present any
+issue for the application.
+
+We need more data to understand the problem directly. We want to drive the popularity
+and adoption of Tendermint and this will mean allowing for chains with more validators.
+We should follow up with users experiencing this issue. We may then want to add
+a series of metrics to the P2P layer to better understand the inefficiencies it produces.
+
+The following metrics can help us understand the sources of latency in the Tendermint P2P stack:
+
+* Number of messages sent and received per second
+* Time of a message spent on the P2P layer send and receive queues
+
+The following metrics exist and should be leveraged in addition to those added:
+
+* Number of peers node's connected to
+* Number of bytes per channel sent and received from each peer
+
+### Sync
+
+#### Block Syncing is slow
+
+Bootstrapping a new node in a network to the height of the rest of the network
+takes longer than users would like. Block sync requires fetching all of the blocks from
+peers and placing them into the local disk for storage. A useful line of inquiry
+is understanding how quickly a perfectly tuned system _could_ fetch all of the state
+over a network so that we understand how much overhead Tendermint actually adds.
+
+The operation is likely to be _incredibly_ dependent on the environment in which
+the node is being run. The factors that will influence syncing include:
+1. Number of peers that a syncing node may fetch from.
+2. Speed of the disk that a validator is writing to.
+3. Speed of the network connection between the different peers that node is
+syncing from.
+
+We should calculate how quickly this operation _could possibly_ complete for common chains and nodes.
+To calculate how quickly this operation could possibly complete, we should assume that
+a node is reading at line-rate of the NIC and writing at the full drive speed to its
+local storage. Comparing this theoretical upper-limit to the actual sync times
+observed by node operators will give us a good point of comparison for understanding
+how much overhead Tendermint incurs.
+
+We should additionally add metrics to the blocksync operation to more clearly pinpoint
+slow operations. The following metrics should be added to the block syncing operation:
+
+* Time to fetch and validate each block
+* Time to execute a block
+* Blocks sync'd per unit time
+
+### Application
+
+Applications performing complex state transitions have the potential to bottleneck
+the Tendermint node.
+
+#### ABCI block delivery could cause slowdown
```

Comment:
This is one of the more important examples of a pattern that seems to recur in TM and the SDK, where we schedule work in parallel that isn't safely concurrent. This results in widespread (ab)use of channels and locks in ways that appear to make things slower (and probably do—but it's hard to prove). 

There are probably some cases where we've been too conservative about synchronization, but the fact that it's hard to understand the code well enough to tell is a red flag.

Another recent case in point is the long discussion/exploration leading up to #6900: In order to tell where it was safe to skip locking, we basically had to wire it up and see what failed, repeatedly. Experimentation is good, but the need to do so much of it to understand the impact of an essential correctness issue worries me.

Anyway, I agree this is a good example, and I suspect there are several other places where poorly-understood (and possibly underspecified) sync discipline is biting us on performance.